Persist snapshot checksum as sibling marker file

npepinpe opened the pull request:
## Description

This PR modifies how we persist snapshot checksum. The checksum file will is not contained in the snapshot directory anymore, but lives next to it. It has the same file name, but with a `.checksum` extension. The contents of the file were not changed and are still just the CRC32C checksum of the directory (note: this was changed in a separate PR).

Validation of the checksum on persisting is done after the snapshot has been moved to its final location. Then, if the checksum matches the expected checksum, the store will persist it as a sibling `.checksum` file. This file acts as a marker file - on start up, any snapshot which does not have this file is considered to have only been partially written and is immediately deleted. This means that for transient/constructed snapshots, we are calculating it twice - once before moving, and once after. This is somewhat safer, but it's also possibly expensive, and should be challenged/discussed.

The PR also comes with some new utilities - one is `FileUtil#deleteFolderIfExists`, which is inspired from the standard library's `Files#deleteIfExists`, and will ignore `NoSuchFileException` both when deleting contained files and the folder itself. For cases such as when we rollback a snapshot, this can be useful since we don't care too much about it. For cases where it would be important to know we did not delete it, we should use the normal `FileUtil#deleteFolder`.

The PR also adds `DirectoryAssert` which provides some convenience methods to assert that a directory contains exactly the expected paths.

## Related issues

closes #6666

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

npepinpe commented:
You don't need to review closely the last commit - it's essentially moving the interface tests to the root package (as the interfaces are in the production code). You can of course challenge this change, but my point is no need to check line by line :slightly_smiling_face:

deepthidevaki commented on snapshot/src/test/java/io/zeebe/snapshots/impl/FileBasedReceivedSnapshotTest.java:
```diff
@@ -125,12 +126,15 @@ public void shouldRemovePreviousSnapshotOnCommit() throws Exception {
     final PersistedSnapshot secondPersistedSnapshot = takePersistedSnapshot(2L);
     final var secondReceivedPersistedSnapshot =
-        receiveSnapshot(secondPersistedSnapshot).persist().join();
+        (FileBasedSnapshot) receiveSnapshot(secondPersistedSnapshot).persist().join();
```
You can change the return type of `receiveSnapshot` instead of casting it here.

npepinpe commented:
FYI, I've rebased against your changes, but only `SnapshotChecksum` had a conflict.

deepthidevaki reviewed (changes requested):
I have just two main comments:
1. Are we now calculating checksum twice when taking? Is that intended?
2. If we are verifying checksum after moving, we should also delete the directory if checksum fails.

npepinpe commented on snapshot/src/main/java/io/zeebe/snapshots/impl/FileBasedSnapshotStore.java:
```diff
@@ -364,10 +390,28 @@ PersistedSnapshot newSnapshot(final FileBasedSnapshotMetadata metadata, final Pa
+    try {
+      // computing the checksum on the final destination also lets us detect any failures during the
+      // copy/move that could occur
+      actualChecksum = SnapshotChecksum.calculate(destination);
```
It was intended, but as I wrote in the description, I would like to challenge it. As you mentioned, it can be expensive if we have a large snapshot. At the same time, it's more or less the only way to verify that the snapshot was moved and kept its integrity. If we think this is safe to do and only calculate it after the move, we could do so as well, but I was unsure about it.

deepthidevaki replied:
We can discuss it later also, as it is not a breaking change. One thing that I discussed with Nico was to only verify the number of files after moving. The reasoning was - The checksum is calculated in the same context as the move, so there is nothing that can happen in between that can corrupt the snapshot. Just to make sure that move is complete, we can compare the number of files.

npepinpe commented on snapshot/src/main/java/io/zeebe/snapshots/impl/FileBasedSnapshotStore.java:
```diff
@@ -364,10 +390,28 @@ PersistedSnapshot newSnapshot(final FileBasedSnapshotMetadata metadata, final Pa
+      // copy/move that could occur
+      actualChecksum = SnapshotChecksum.calculate(destination);
+      if (actualChecksum != expectedChecksum) {
+        throw new InvalidSnapshotChecksum(directory, expectedChecksum, actualChecksum);
```
Good catch! Forgot to roll back :+1:

npepinpe commented:
I addressed all comments, the last remaining is the one about doing the calculation twice on constructed snapshots - once before moving, once after moving. I figured it's safer (for example, if you're on network storage or things where moving files may be problematic), but it is much more expensive with large state.

We can do this after the code freeze and release it for 1.1  however, or as a patch release, as it wouldn't break backwards compat. Let me know what you think.

deepthidevaki reviewed (approved):
:+1:

npepinpe commented:
bors r+

npepinpe commented:
bors retry

npepinpe commented:
bors retry
